StereoKit/Framework/Steppers.cs

```diff
@@ -1,74 +1,85 @@
+		enum   ActionType { Add, Remove }
+		struct StepperAction
 		{
-			_steppers.ForEach(s => s.Shutdown());
-			_steppers.Clear();
+			public IStepper   stepper;
+			public ActionType type;
+			public StepperAction(IStepper stepper, ActionType type) { this.stepper = stepper; this.type = type; }
 		}
 
+		List           <IStepper>      _steppers = new List<IStepper>();
+		ConcurrentQueue<StepperAction> _actions  = new ConcurrentQueue<StepperAction>();
+
+		// Add steppers via the threadsafe action queue
 		public T Add<T>() where T : IStepper
 		{ 
-			T inst = Activator.CreateInstance<T>(); 
-			Add(inst); 
+			T inst = Activator.CreateInstance<T>();
+			_actions.Enqueue(new StepperAction(inst, ActionType.Add));
 			return inst;
 		}
 		public object Add(Type type)
 		{
 			IStepper inst = Activator.CreateInstance(type) as IStepper;
 			if (inst == null) return null;
-			Add(inst);
+			_actions.Enqueue(new StepperAction(inst, ActionType.Add));
 			return inst;
 		}
 		public T Add<T>(T stepper) where T:IStepper 
 		{
-			// Initialize the stepper! If called from a separate thread, this
-			// needs to happen first, otherwise there's a chance the Step may
-			// happen before Initialize
-			if (SK.IsInitialized)
-				stepper.Initialize();
-
-			// Add the stepper to the list
-			_steppers.Add(stepper);
-
+			_actions.Enqueue(new StepperAction(stepper, ActionType.Add));
 			return stepper;
 		}
-		public void InitializeSteppers()
-		{
-			// This should only be called for steppers that were added before
-			// initialization!
-			System.Diagnostics.Debug.Assert(!SK.IsInitialized);
-			_steppers.ForEach(s => s.Initialize());
-		}
 
+		// Remove steppers via the threadsafe action queue
 		public void Remove<T>() => Remove(typeof(T));
 		public void Remove(Type type)
 		{
-			_steppers.RemoveAll(s => {
-				bool remove = type.IsAssignableFrom(s.GetType());
-				if (remove) s.Shutdown();
-				return remove;
-			});
+			foreach(IStepper stepper in _steppers)
+			{
+				if (type.IsAssignableFrom(stepper.GetType()))
+					_actions.Enqueue(new StepperAction(stepper, ActionType.Remove));
+			}
 		}
 		public void Remove(IStepper stepper)
 		{
-			if (_steppers.Remove(stepper))
-				stepper.Shutdown();
+			_actions.Enqueue(new StepperAction(stepper, ActionType.Remove));
 		}
 
-
 		public void Step()
 		{
-			for (int i = 0; i < _steppers.Count; i++)
+			// Execute all stepper actions on the main thread in the order they
+			// were given.
+			while (_actions.TryDequeue(out StepperAction action))
 			{
-				_steppers[i].Step();
+				switch (action.type)
+				{
+					case ActionType.Add:
+						action.stepper.Initialize();
+						_steppers.Add(action.stepper);
+						break;
+					case ActionType.Remove:
+						_steppers.Remove(action.stepper);
+						action.stepper.Shutdown();
+						break;
+				}
 			}
+
+			foreach (IStepper stepper in _steppers)
+				stepper.Step();
+		}
+
+		public void Shutdown()
+		{
+			_actions = new ConcurrentQueue<StepperAction>();
```

Comment:
If shutdown is called early, we should give some indicator to halt the `Step` or we get the following exception:
![image](https://user-images.githubusercontent.com/32556981/205721826-7647fa20-4447-4090-ad04-8308350e1bc1.png)
I'll come up with an idea and push to this branch, if that's okay

Reply:
Oh interesting! I think I'd consider this to be an invalid use of `SK.Shutdown`, and I may want to add some guards to prevent it from happening? I'll probably throw an exception if called while SK is still running. `SK.Quit` is the correct function to call in this case.

Reply:
`SK.Quit` works perfectly with this test case. Cool, so `Quit` must be called to step through the last frame, then it shutdowns. An exception would be great!

Reply:
Exception added :)